### PR TITLE
Adding Aggregations

### PR DESCRIPTION
**Pull request by pdet**
This PR implements the following aggr functions:
The only function missing from #1125 is the any (alias for first) because the Postgres parser reserves it to be used with subqueries. Not sure how (or even if I should dare to) change the parser.
* corr
* product
* bool_and
* bool_or
* approx_count_distinct (Had to bring hyperloglog from redis back)
* regr_avgx
* regr_avgy
* regr_count
* regr_intercept
* regr_r2
* regr_slope
* regr_sxx
* regr_sxy
* regr_syy
* mode
* argMin
* argMax
* variance (alias for var_samp)

### CODE REVIEW

**Comment by hannesmuehleisen** [comment]
This looks really good! 

I have three minor comments:

1. I'm cool with many of those functions only being defined for `DOUBLE`, but for `APPROX_COUNT_DISTINCT ` you do define more types. Would that not also be applicable for things like `PRODUCT`? 
2. What about DECIMAL types, they seem not handled. They are likely auto-cast to DOUBLE, but that would be counter-productive in many cases I think
3. Could you add some tests grouped aggregates perhaps?

**Comment by pdet** [comment]
The reason I only went for double only in the product was to diminish the overflow problem, if adding all types this would need to be checked and handled for each multiplication I think.
Good point, I'll extend them to Decimal and add grouped test cases for those that do not have that.
Any suggestions about "any"?

File: src/function/aggregate/distributive/approx_count.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+		                                                                                LogicalType::BIGINT);
+
+	default:
+		throw NotImplementedException("Unimplemented approximate_count aggregate");
+	}
+}
+
+void ApproxCountDistinctFun::RegisterFunction(BuiltinFunctions &set) {
+
+	AggregateFunctionSet approx_count("approx_count_distinct");
```
Comment (Mytherin) [inline src/function/aggregate/distributive/approx_count.cpp:104]: What about strings, dates, timestamps?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR, this is the bees knees. Some comments, mostly suggestions for extra tests:

File: test/sql/aggregate/aggregates/test_mode.test [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+statement ok
+insert into names values ('pedro'), ('pedro'), ('pedro'),('hannes'),('hannes'),('mark'),(null);
+
+query I
+select mode(name) from names;
+----
+pedro
+
+require vector_size 512
```
[the remark this answers is not in the document]
Comment (pdet) [reply test/sql/aggregate/aggregates/test_mode.test:53]: window function on the test

[the remark this answers is not in the document]
Comment (Mytherin) [reply test/sql/aggregate/aggregates/test_mode.test:53]: Ah, excellent point. Can you add window function tests for all the aggregates that don't have them yet as well?

File: test/sql/aggregate/aggregates/test_approximate_distinct_count.test [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+# group: [aggregates]
+
+statement error
+select approx_count_distinct(*)
+
+statement ok
+select approx_count_distinct(1)
+
+statement ok
+select approx_count_distinct(NULL)
```
[the remark this answers is not in the document]
Comment (pdet) [reply test/sql/aggregate/aggregates/test_approximate_distinct_count.test:12]: hyperloglog doesnt always return the same result, but I aguess I can add something similar to what I did for the approximate queries with some error %

[the remark this answers is not in the document]
Comment (Mytherin) [reply test/sql/aggregate/aggregates/test_approximate_distinct_count.test:12]: Does hyperloglog involve randomness? If not it should always return the same result with the same input set, no?

[the remark this answers is not in the document]
Comment (pdet) [reply test/sql/aggregate/aggregates/test_approximate_distinct_count.test:12]: Yeah you are right, it is deterministic, will add those.

[the remark this answers is not in the document]
Comment (Mytherin) [reply test/sql/aggregate/aggregates/test_approximate_distinct_count.test:12]: Not 100% sure but I think it works similar to bloom filters in that the order in which the input comes doesn’t matter, so the result should also be consistent when processing in parallel.

File: test/sql/aggregate/aggregates/test_bool.test [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+# Corner Cases
+statement error
+select bool_or(0)
+
+statement error
+select bool_and(0)
+
+query I
+select bool_or(NULL)
```
Comment (Mytherin) [inline test/sql/aggregate/aggregates/test_bool.test:13]: Can you add a test of bool_or and bool_and with a mix of NULL values and actual values?

File: test/sql/aggregate/aggregates/test_approximate_distinct_count.test [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+8	4
+
+statement ok
+PRAGMA threads=4
+
+statement ok
+PRAGMA force_parallelism
+
+query III
+SELECT COUNT( a),approx_count_distinct(a),approx_count_distinct(b) from t
```
Comment (Mytherin) [inline test/sql/aggregate/aggregates/test_approximate_distinct_count.test:99]: t should be a bit bigger to meaningfully test the parallelism (>10K elements), also please move the parallel tests to a separate file in the parallel directory

File: src/function/aggregate/distributive/approx_count.cpp [hunk 6]
```diff
@@ -86,10 +86,8 @@ struct ApproxCountDistinctFunctionString : ApproxCountDistinctFunctionBase {
 		if (nullmask[idx]) {
 			return;
 		}
-		string value = input[idx].GetString();
-		//! FIXME: Hash this in the future
-		auto size = value.size() > 8 ? 8 : value.size();
-		state->log->Add((uint8_t *)&value, size);
+		auto str_hash = std::hash<std::string> {}(input[idx].GetString());
```
Comment (Mytherin) [inline src/function/aggregate/distributive/approx_count.cpp:89]: Can you use our own hash instead of std::hash? This is not very efficient as it constructs a std::string object which involves memory allocation and copying the string.
